packages/flutter/lib/src/services/text_formatter.dart

```diff
@@ -363,14 +365,30 @@ class LengthLimitingTextInputFormatter extends TextInputFormatter {
   /// composing is not allowed.
   final int? maxLength;
 
+  /// Whether the [maxLength] constraint will be applied when the editing value
+  /// is composing.
+  ///
+  /// Defaults to false.
+  final bool composingMaxLengthEnforced;
+
   /// Truncate the given TextEditingValue to maxLength user-perceived
   /// characters.
   ///
+  /// ### Composing text behaviors
+  ///
+  /// If [composingMaxLengthEnforced] and [keepComposing] are both true,
```

Comment:
```suggestion
The composing range of [value] will be set to [TextRange.empty] if [keepComposing] is false. Otherwise only the part that exceeds [maxLength] will be truncated.
```

Reply:
Alternatively you can just make this method always accounts a `TextEditingValue` with truncated composing region, and change the composing range to (-1, -1) in `formatEditUpdate`. Up to you.